source/MRMesh/MRObject.h

```diff
@@ -345,4 +362,29 @@ std::shared_ptr<const T> Object::find( const std::string_view & name ) const
 
 /// \}
 
+struct KeyObjectModel
+{
+    const Object* object = nullptr;
+};
+
+// return true if model data of two Objects are equal
+inline bool operator==( const KeyObjectModel& a, const KeyObjectModel& b )
+{
+    if ( !a.object || !b.object )
+        return a.object == b.object;
+    return a.object->isModelEqual( *b.object );
+}
+
+// return hash for Object model data
+struct KeyObjectModelHasher
+{
+    MRMESH_API std::size_t operator()( const KeyObjectModel& k ) const;
+};
+
+struct LinkData
+{
+    std::string link; // relative path (link) to shared model file
+    bool serialized = false; // set true after first serialization to avoid other model copies
+};
```

Comment:
can we use `const Object*` instead of this structure? (so it will be single MapType for both serialization and deserialization)
```c++
auto it = map.find( this );
if ( it != map.end() )
    // add link from it->second->createLink()
else
   // serialize as not-shared model
if ( it->second == this )
    // serialize as shared model
```